core/src/main/scala/io/snappydata/SnappyTableStatsProviderService.scala

```diff
@@ -215,55 +221,265 @@ object SnappyEmbeddedTableStatsProviderService extends TableStatsProviderService
+      if ((localPrimaries ne null) && localPrimaries.size() > 0) {
+        val doRollover = new Predicate[BucketRegion] {
+          private val minModTime = pr.getCache.cacheTimeMillis() - delayMillis
+
+          override def test(br: BucketRegion): Boolean = {
+            br.getLastModifiedTime <= minModTime && br.getRegionSize >= minSizeForRollover(pr)
+          }
+        }
+        val rolloverBuckets = localPrimaries.asScala.filter(
+          br => doRollover.test(br) && !br.columnBatchFlushLock.isWriteLocked)
+        // enqueue a job to roll over required row buffers into column table
+        // (each bucket will perform a last minute check before rollover inside lock)
+        if (rolloverBuckets.nonEmpty) {
+          implicit val executionContext = Utils.executionContext(pr.getGemFireCache)
+          Future {
+            try {
+              rolloverBuckets.foreach(bucket => withExceptionHandling(
+                bucket.createAndInsertColumnBatch(null, true, doRollover)))
+            } finally {
+              rolloverTasks.remove(pr)
+            }
+          }
+        } else null
+      } else null
+    }
+  }
+
+  /**
+   * Merge multiple column batches that are small in size in a bucket.
+   * These can get created due to a small "tail" in bulk imports (large enough
+   * to exceed minimal size that would have pushed them into row buffers),
+   * or a time-based flush that tolerates small sized column batches due to
+   * [[rolloverRowBuffersTask]] or a forced flush of even smaller size for sample tables.
+   */
+  private def mergeSmallColumnBatches(pr: PartitionedRegion, container: GemFireContainer,
+      metaData: ExternalTableMetaData, smallBatches: mutable.ArrayBuffer[BucketRegion]): Unit = {
+    mergeTasks.computeIfAbsent(pr, new JFunction[PartitionedRegion, Future[Unit]] {
+      override def apply(pr: PartitionedRegion): Future[Unit] = {
+        logInfo(s"Found small batches in ${pr.getName}: ${smallBatches.map(_.getId)}")
+        val cache = pr.getGemFireCache
+        implicit val executionContext = Utils.executionContext(cache)
+        Future(withExceptionHandling({
+          val tableName = container.getQualifiedTableName
+          val schema = metaData.schema.asInstanceOf[StructType]
+          val maxDeltaRows = pr.getColumnMaxDeltaRows
+          val compileKey = tableName.concat(".MERGE_SMALL_BATCHES")
+          val gen = CodeGeneration.compileCode(compileKey, schema.fields, () => {
+            val schemaAttrs = Utils.schemaAttributes(schema)
+            val tableScan = ColumnTableScan(schemaAttrs, dataRDD = null,
+              otherRDDs = Nil, numBuckets = -1, partitionColumns = Nil,
+              partitionColumnAliases = Nil, baseRelation = null, schema, allFilters = Nil,
+              schemaAttrs, caseSensitive = true)
+            // reduce min delta row size to avoid going through rolloverRowBuffers again
+            val insertPlan = ColumnInsertExec(tableScan, Nil, Nil,
```

Comment:
Can you please explain here, maybe in the comment section how this will handle updated columns and deleted columns?

Reply:
The ColumnBatchIterator is passed the stats row entry. Rest all the columns, including delta/delete are looked up by the iterator (see ColumnFormatStatsIterator.getColumnValue) when the generated code asks for them.

So this will be same as iterating batches in ColumnTableScan that will return merged entries with deltas/deletes applied. The ColumnInsert is tied to output of this hence will create a combined merged batch.

Note that merging of deltas into main batch (when they become large enough) or deletes into main batch (when large number of entries are deleted) will be handled separately. This does not depend on main batch being small rather deltas being large. That merge needs to be done in the operation thread itself (that has created the last delta causing it to grow large). Right now only the case where all entries are deleted is handled.